src/sensors/SensorManager.cpp

```diff
@@ -33,109 +33,171 @@
-#elif SECOND_IMU == IMU_BNO055
-                    m_Sensor2 = new BNO055Sensor(sensorID, secondIMUAddress, SECOND_IMU_ROTATION);
-#elif SECOND_IMU == IMU_MPU9250
-                    m_Sensor2 = new MPU9250Sensor(sensorID, secondIMUAddress, SECOND_IMU_ROTATION);
-#elif SECOND_IMU == IMU_BMI160
-                    m_Sensor2 = new BMI160Sensor(sensorID, secondIMUAddress, SECOND_IMU_ROTATION);
-#elif SECOND_IMU == IMU_MPU6500 || SECOND_IMU == IMU_MPU6050
-                    m_Sensor2 = new MPU6050Sensor(sensorID, SECOND_IMU, secondIMUAddress, SECOND_IMU_ROTATION);
-#elif SECOND_IMU == IMU_ICM20948
-                    m_Sensor2 = new ICM20948Sensor(sensorID, secondIMUAddress, SECOND_IMU_ROTATION);
-#endif
-                }
+        void SensorManager::swapI2C(uint8_t sclPin, uint8_t sdaPin)
+        {
+            if (sclPin != activeSCL || sdaPin != activeSDA || !running) {
+                Wire.flush();
+                #if ESP32 
+                    if (running) {}
+                    else {
+                        // Reset HWI2C to avoid being affected by I2CBUS reset
+                        Wire.end();
+                    }
+                    // Disconnect pins from HWI2C
+                    pinMode(activeSCL, INPUT);
+                    pinMode(activeSDA, INPUT);
+                    
+                    if (running) {
+                        i2c_set_pin(I2C_NUM_0, sdaPin, sclPin, false, false, I2C_MODE_MASTER);
+                    } else {
+                        Wire.begin(static_cast<int>(sdaPin), static_cast<int>(sclPin), I2C_SPEED); 
+                        Wire.setTimeOut(150);
+                    }
+                #else
+                    Wire.begin(static_cast<int>(sdaPin), static_cast<int>(sclPin));
+                #endif
+
+                activeSCL = sclPin;
+                activeSDA = sdaPin;
+            }
+        }
 
-                m_Sensor2->motionSetup();
+        void SensorManager::setup()
+        {
+            running = false;
+            activeSCL = PIN_IMU_SCL;
+            activeSDA = PIN_IMU_SDA;
+
+            uint8_t sensorID = 0;
+            uint8_t activeSensorCount = 0;
+#define IMU_DESC_ENTRY(...)                                          \
+            {                                                        \
+                Sensor* sensor = buildSensor(sensorID, __VA_ARGS__); \
+                m_Sensors[sensorID] = sensor;                        \
+                sensorID++;                                          \
+                if (sensor->isWorking()) {                           \
+                    m_Logger.info("Sensor %d configured", sensorID); \
+                    activeSensorCount++;                             \
+                }                                                    \
+            }
+            // Apply descriptor list and expand to entrys
+            IMU_DESC_LIST;
+
+#undef IMU_DESC_ENTRY
```

Comment:
This feels a bit hacky, but I guess it works

Reply:
Yes it's kind of preprocessor spells here. That's also my first time seeing it used in an MCU project.